[Java][jersey2] Fix generated client code for oneOf models if datatype includes arrays

B4ckslash opened the pull request:
## Description
#### Changes
As has already been described in #17745, the code generated for Java with Jersey2 for a oneOf-model can be invalid if one of the possible types is of `array` type. This is presumably due to the fact that #13897 only fixed the templates for okhttp-gson, but not other serialization libraries. This PR updates the `oneOf_model` template for jersey2 as well as add a sample which generates a mixed oneOf model using jersey2 so as to catch regressions on this in future.

Beyond that, I've added a mustache lambda to `AbstractJavaCodegen` which replaces angle brackets, spaces and commas with nothing. This is done so that the data type of any parametrized type may be used in identifiers (e.g. method names of getters) to prevent clashing method signatures. This example used to fail compilation due to clashing erasures:
```java
public List<String> getList() {...}
public List<List<Integer>> getList() {...}
```
whereas now, one can generate something like:
```java
public List<String> getListString() {...}
public List<List<Integer>> getListListInteger {...}
```
This also has the added benefit of automatically changing the method name should the return type ever change, forcing a user to think about handling the changed datatype, which would be impossible if it were a simple `public List<?> getList()`. 
Another benefit is the ability to use the actual generic type that the user can expect instead of just giving them a wildcarded List.
#### Possible future work
* I don't use jersey3, which probably suffers from the same problem.
* A commment in #17745 mentions that C# has a similar issue. Perhaps a similar fix would be applicable there?

<!-- Please check the completed items below -->
## PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.1.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@bbdouglas @sreeshas @jfiala @lukoyanov  @cbornet @jeff9finger  @karismann  @Zomzog  @lwlee2608 @martin-mfg

wing328 commented:
thanks for the PR. the CI tests failed: https://circleci.com/gh/OpenAPITools/openapi-generator/90090

can you please take a look when you've time?

B4ckslash commented:
@wing328 I think the CI ought to work now. However, looking through the templates for other Java libraries (esp. Gson), I think I may have found a better way to go about this fix with only adjusting the templates.
In any case, I will add a sample and/or test case to properly test this behavior before re-doing the fix.

B4ckslash commented:
@wing328 ok I'm done. Perhaps someone from the Java experts can look over this now.

wing328 commented on bin/configs/java-jersey2-8-oneOfMixed.yaml:
```diff
@@ -0,0 +1,7 @@
+generatorName: java
+outputDir: samples/client/others/java/jersey2-oneOf-Mixed
```
can you please add the new sample folders to `.github/workflows/samples-java-client-jdk11.yaml` so that CI will test it moving forward?

B4ckslash replied:
Done.

wing328 commented:
when you've time, can you please PM me via Slack when you've time?

I may have other ways to handle array in oneOf/anyOf schema bette.r

B4ckslash commented:
Yeah I can do that, if you tell me where and how to find you. I've not used slack for a long while.

wing328 commented:
please join https://join.slack.com/t/openapi-generator/shared_invite/zt-12jxxd7p2-XUeQM~4pzsU9x~eGLQqX2g and search for "William Cheng"

B4ckslash commented:
Alright so as per our Slack chat, I've changed the following (and updated the original description where applicable):

- Primitives are handled more explicitly in the template, using the appropriate flags in the corresponding property. This blows up the template a bit, but reduces generated code a fair amount (e.g. completely removing the runtime class check when the property is not primitive)
- The pre-escaped data type got replaced with a mustache lambda which strips angle brackets, commas, and spaces

Assuming the tests pass, this PR is finally ready to be merged :partying_face:

wing328 commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:
```diff
@@ -8463,11 +8468,16 @@ public Set<String> getOpenAPIGeneratorIgnoreList() {
         return openapiGeneratorIgnoreList;
     }
 
+    @Override
```
can you describe what `isTypeErasedGenerics` does with a line or two and I'll add a docstring to the method below in my upcoming PR?

B4ckslash replied:
Sorry for the late reply. `isTypeErasedGenerics`  is meant to tell the code generator whether the target language is erasing the generic parameter(s) at runtime, which has implications for handling the possible types in things like the constructor or `instanceof` checks. As an example, if the language has type erased generics, the compiler throws an error when you do something like 
```java
public Constructor(final List<String> args) {...}

public Constructor(final List<Integer> args) {...}
```
because both the signatures erase to the same signature `Constructor(final List args)` which is not allowed. Therefore you need a way to check whether the code generation can include the full parametrized type `List<String>` and do so potentially multiple times, or if it has to only use it once as the base type `List`.

wing328 replied:
what about using ` {{^vendorExtensions.x-duplicated-data-type}} ...  {{/vendorExtensions.x-duplicated-data-type}}`  instead?

https://github.com/B4ckslash/openapi-generator/blob/3702fa2a7db0fbfac3007c9e2f6eb095e0992c4e/modules/openapi-generator/src/main/resources/Java/libraries/jersey2/oneof_model.mustache#L194C1-L195C1

B4ckslash replied:
Yes I use that as well, but I have to determine on the codegen side whether that is even applicable. In languages that do not erase their generics, `List<String>` and `List<Integer>` are different data types, so in those languages using `x-duplicated-data-type` makes no sense. Depending on whether `isTypeErasedGenerics` is true, I set the `x-duplicated-data-type` flag for types affected by type erasure (if there are multiple variants of it).

wing328 commented:
FYI. I've filed #18274 to sync the templates between jersey2 and jersey3

B4ckslash commented:
I can have a look at it in a few days time I'm sure.
